ins-node/src/wifi_module/wifi_module.ino

```diff
@@ -1,10 +1,156 @@
+const int8_t CONNECTION_RETRIES = 20; // The maximum amount of wifi connection attempts
+const uint8_t DEFAULT_TRANSMISSION_SIZE = 10;
+uint8_t currentTransmissionSize = DEFAULT_TRANSMISSION_SIZE;
+
+const uint8_t TX_PIN = 10;
+const uint8_t RX_PIN = 9;
+
+/**
+   Gets the network context (i.e. MAC address and RSSI value) of the specified
+   network item.
+   @param networkItem The network item index of which to fetch the context
+   @return            A tuple of (MAC address, RSSI)
+*/
+std::pair <String, int32_t> getNetworkContext(int8_t networkItem) {
+  std::pair <String, int32_t> context (wifi.BSSIDstr(networkItem), wifi.RSSI(networkItem));
+  return context;
+}
+
+/**
+   Measures the RSSI values for the various access points of the specified SSID
+   and returns them along the MAC address of the respective access point.
+   @return  The MAC addresses with their respective RSSI values
+*/
+std::vector<std::pair <String, int32_t>> getDatapoints() {
+  std::vector<std::pair <String, int32_t>> datapoints;
+  // Scan the surrounding networks and get how many were found
+  auto networksFound = wifi.scanNetworks();
+  for (auto i = 0; i < networksFound; ++i) {
+    if (wifi.SSID(i) == positioningSSID) {
+      datapoints.push_back(getNetworkContext(i));
+    }
+  }
+
+  return datapoints;
+}
+
+/**
+   Transmits the supplied data (if present) to the server
+   @param  datapoints The MAC addresses and the RSSI values we collected
+   @return            Whether transmission succesfull
+*/
+bool transmitData(std::vector<std::pair <String, int32_t>> datapoints) {
+  // TO-DO: Notify the server that we could not gather any datapoints for
+  // corrective actions to be made before sleeping.
+  if (datapoints.empty()) {
+    return false;
+  }
+  // Establish a TCP connection to the server
+  WiFiClient client;
+  if (!client.connect(SERVER_IP, HTTP_PORT)) {
+    return false;
+  }
+
+  // Compose a request for the server according to the agreed format:
+  // /set_rssi/device_id/mac_addr1/rssi1/mac_addr2/ ... /mac_addr10/:rssi10/
+  String request = "/set_rssi/";
+  request += DEVICE_ID;
+  request += "/";
+  for (auto datapoint : datapoints) {
+    // We don't put everything in the same expression since std::string
+    // (used in testing) is not exactly the same as Arduino String. Therefore
+    // we go with this way of concatinating the string which is compatible with
+    // both implementations.
+    request += datapoint.first; // The MAC address
+    request += "/";
+    request += datapoint.second; // The RSSI value
+    request += "/";
+  }
+
+  // Send the TCP request to the rest server
+  client.print("POST " + request + " HTTP/1.1\r\n" +
+               "Host: " + SERVER_IP + "\r\n" +
+               "Connection: close\r\n\r\n");
+
+  // Verify whether the request timed out
+  bool successfulRequest = true;
+  const unsigned long TIMEOUT = millis() + 5000;
+  while (client.available() == 0) {
```

Comment:
I see we are polling on the client to see when it becomes available again after the http request has been sent. While this may help determine busy-state implying timeout, are there other actions that can cause the client to be unavailable? Should we handle cases where we get negative response from INS service?
why not included in header file?

Also, in a case of timeout, do we know the expected behavior of the client?  e.g. when does it becomes available again?

Reply:
> While this may help determine busy-state implying timeout, are there other actions that can cause the client to be unavailable? Should we handle cases where we get negative response from INS service?

Yes, we should definitely handle responses from the server in the future.

> why not included in header file?
Not sure what you mean.

> Also, in a case of timeout, do we know the expected behavior of the client? e.g. when does it becomes available again?

If I understood you correctly, I have a `TO-DO` on line 149 about this. We would need to discuss how to handle these errors/cases. Currently it will stop trying to send and eventually go to sleep.

Reply:
>why not included in header file?
>Not sure what you mean.

Forget this, that was a copy paste by error

>If I understood you correctly, I have a TO-DO on line 149 about this. We would need to discuss how to handle these errors/cases. Currently it will stop trying to send and eventually go to sleep.

OK, must have overlooked the TO-DO line.